### PR TITLE
[CLIPPER-146] Bug fixes, logging improvements, documentation for python function deployment

### PR DESCRIPTION
**Pull request by nishadsingh1**
Expanded python_container_conda_deps to include more packages. Specifically, it now includes those that are necessary to run through the tutorial (since many first time users will probably just use the same conda environment). This choice will decrease the container startup time in this case but is less resource-efficient when users are working with a more minimal conda environment, so I’m happy to reverse it.

Changed python_container_entry.sh so that it doesn’t break

Added some more logging for clarity and stderr/stdout flushes at appropriate points so that it’s easy to diagnose the status of a  python function container’s initialization

Added a (very basic) guide on deploying arbitrary python prediction functions. Will have to add to/edit this more.

### CODE REVIEW

**Comment by nishadsingh1** [comment]
Parts of the documentation assume that the changes in https://github.com/ucbrise/clipper/pull/130 make it through

**Comment by nishadsingh1** [comment]
This documentation in this PR assumes that the changes in #147 (#147 replaced #130) go through

File: containers/python/python_container_conda_deps.txt [hunk 1]
```diff
@@ -1,9 +1,82 @@
+backports
```
Comment (dcrankshaw) [inline containers/python/python_container_conda_deps.txt:1]: How did you generate this list?

File: containers/python/python_container_entry.sh [hunk 2]
```diff
@@ -1,5 +1,4 @@
 #!/usr/bin/env sh
 
 conda env create -f=/model/environment.yml -n container-env
-source activate container-env
-python /container/python_container.py
+/bin/bash -c "source activate container-env && exec python /container/python_container.py"
```
Comment (dcrankshaw) [inline containers/python/python_container_entry.sh:4]: Why `exec python` and not just `python`?

File: containers/python_function_deployment.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
+# Python prediction function deployment
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:1]: This should probably go in `containers/python`

File: containers/python_function_deployment.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
+# Python prediction function deployment
+Clipper supports a simplified process for deploying arbitrary Python prediction functions as model-containers. 
+
+The Clipper management library supports this functionality through `deploy_prediction_func`, which consumes a the function you want to deploy. All dependencies for the function must be installed with Anaconda or Pip. `deploy_prediction_func` must be called from within an [Anaconda environment](https://conda.io/docs/using/envs.html) that tracks all such dependencies.
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:4]: Typo: s/consumes a the/consumes a

File: containers/python_function_deployment.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
+# Python prediction function deployment
+Clipper supports a simplified process for deploying arbitrary Python prediction functions as model-containers. 
+
+The Clipper management library supports this functionality through `deploy_prediction_func`, which consumes a the function you want to deploy. All dependencies for the function must be installed with Anaconda or Pip. `deploy_prediction_func` must be called from within an [Anaconda environment](https://conda.io/docs/using/envs.html) that tracks all such dependencies.
+
+## Defining your prediction function
+Prediction functions should take a list of inputs of and return a Python or numpy array of predictions. Internally, the function is free to do anything so long as the function's closure captures its full state.
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:7]: typo: s/list of inputs of/list of inputs

Comment (dcrankshaw) [reply containers/python_function_deployment.md:7]: Also, specify that the function should take a list of inputs of the specified type (e.g doubles). Also, models return strings not floats. So the function should return a list of JSON strings the same length as the list of inputs, where the output at index `i` of the list is the prediction for input at index `i`.

Also add a note that any transient state (e.g. a database or network connection) will not be captured correctly.

File: containers/python_function_deployment.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+      means = np.mean(xs, axis=0)
+      return xs - means
+
+    centered_xs = center(xs)
+    model = sklearn.linear_model.LogisticRegression()
+    model.fit(centered_xs, ys)
+
+    def centered_predict(inputs):
+      centered_inputs = center(inputs)
+      return model.predict(centered_inputs)
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:22]: Wrap this code block in triple-backticks to get Python syntax highlighting:

````
```py
# code goes here...
```
````

File: containers/python_function_deployment.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+## Requesting deployment of  your Python function
+Deploying this function to Clipper doesn't require writing any container code. All that's needed is a call to `deploy_predict_function` from within an Anaconda environment that has necessary conda and pip packages to run the function.
+
+    clipper.deploy_predict_function(
+      "example_model",
+      1,
+      centered_predict,
+      ["example"],
+      "doubles",
+      num_containers=1)
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:33]: triple backticks for syntax highlighting

File: containers/python_function_deployment.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    Your conda dependencies are unsatisfiable (see error text below). Please resolve these issues and call `deploy_predict_func` again.
+    UnsatisfiableError: The following specifications were found to be in conflict:
+      - appnope 0.1.0 py27_0
+      - scikit-learn 0.17.1 np110py27_0 -> mkl 11.3.1
+      - scikit-learn 0.17.1 np110py27_0 -> numpy 1.10*
+
+
+###2. Checking for package existence in linux channels
+
+If none of your packages have conflicting dependencies, the management library will check for the existence of your packages in the linux-64 conda channel (the channel the container will use). If any such packages exist, they will be skipped during package installation on the container. You should see a message like this:
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:52]: Typo: "If any such packages exist, they will be skipped..." I think you meant something like "If any your packages cannot be found..."

File: containers/python_function_deployment.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+### 3. Starting up the model in the newly created Anaconda environment
+
+In this stage, the container attempts to load in the parameters you have provided it and connect to Clipper accordingly. The output of `docker logs` should show:
+
+```
+Starting PythonContainer container
+Connecting to Clipper with <port information>
+```
+
+### 4. Loading your prediction function
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:120]: Can you double check that 3 and 4 happen in this order? I think that we load the model before connecting to Clipper, but I could be wrong.

File: containers/python_function_deployment.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+Serving predictions for <input_type> input type.
+```
+
+At this point, your prediction function and its environment have successfully been loaded and are ready to be queried!
+
+## Errors you can encounter
+
+There are several problems you can encounter in deploying your Python function to Clipper. This section will walk you through how you can identify and resolve them.
+
+If you run into issues not listed here, please file an issue and we will attempt to find a generalizable approach to diagnosing and solving it.
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:140]: Specify to file a GitHub issue

File: containers/python_function_deployment.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+### Remote during container initialization
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:142]: What does this section title mean? Specifically, what is remote here?

File: containers/python_function_deployment.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+    Supplied environment details
+    Serialized and supplied predict function
+    Published model to Clipper
+    Found clipper/python-container in Docker hub
+    Copied model data to host
+    True
+
+## Initializing your model-container
+Calling `deploy_prediction_func` spins up a container and loads it with the necessary info to create your model. However, it may take some time before the model-container is ready to serve predictions: it still has to install and load dependencies and deserialize your prediction function.
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:76]: Specify that "some time" could be several minutes.

File: containers/python_function_deployment.md [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+This can occur if:
+ 
+* Your function makes use of a pip package that your conda environment wasn't tracking. See the "Some pip packages weren't installed" section above.
+* Your function makes use of a conda package that doesn't exist in the container's conda channel and whose installation was skipped. If any packages in your local conda environment were skipped for installation on the container, you should have been notified of them in a warning message after calling `deploy_predict_function`.
+* Conda and pip both don’t have a dependency being used by the function. Unfortunately, Clipper's Python function deployment feature only supports conda and pip packages for now. Usage of modules not within these distributions is incompatible with this deployment process.
+
+**Container crash: prediction function has some other runtime error (index out of bounds, divide by zero, incorrect operand types for a function, etc**
+
+Make sure that you test your function locally before deploying it. Remember that it should take as input a list of datapoints (numpy arrays or native Python lists) with entries of type <input_type> and shoud output a list with a classification for each datapoint.
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:172]: typo s/shoud/should. Remind them once again that the output should be a list of strings.

File: containers/python_function_deployment.md [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+
+**Container crash: function makes use of something that’s not captured in the function closure**
+
+Your function can't use objects or connections whose states aren't captured in the function closure. An example of this would be a database connection that is configured/initialized outside of the function.
+
+**Container logs indicate no issues but the Python function's predictions aren't getting served**
+First, make sure that your Python function is actually getting called. One way to do this is to check the container's logs for print statements in your Python function.
+
+If you can confirm that your function is indeed getting called, then thereare a couple of things that could prevent its prediction from getting served:
```
Comment (dcrankshaw) [inline containers/python_function_deployment.md:181]: Typo s/thereare/there are

Rephrase to: "then there are a few reasons that could prevent predictions from being served"

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
This looks great!

**Comment by dcrankshaw** [comment]
I'm going to combine this PR with #147 because it seems like some of the changes here necessary to fix bugs in #147

**Comment by dcrankshaw** [comment]
Combined with #147

**Pull request closed without merging**